Fix -no-shared-libs support

dra27 opened the pull request:
While trying to get Cygwin64 CI to do something useful, I came across various `-no-shared-libs` limitations.

 - #1063 didn't update the `nodynlink.ml` file, which is done here
 - #1691 failed to update `utils/config.mli`, which is now done here
 - A small bug in the error message for `ocamlc -compat-32` when running with `-custom` is included (the name of the temporary bytecode executable in `/tmp` would be displayed instead of the _final_ executable name)
 - The caml-tex test requires a new `no-shared-libraries` ocamltest action. It would also be possible to do something more intelligent in `ocamlrun` here (i.e. have `ocamlrun ./something-compiled-with-custom` realise that it doesn't need interpreting)
 - When shared library support is not available, ocamltest should link with `-custom` if any of the libraries (e.g. `unix.cma`) have C stubs. This is done by querying the cma files directly.
 - A similar check is made for running the toplevel with a library. In theory, this could build a custom toplevel but as it only affects the `unboxed-primitive-args` test, I opted just to skip it.

dra27 committed:
Expose Config.supports_shared_libraries

Detail added to the printed output in GPR#1691, but not exposed from the
module itself.

dra27 committed:
Fix support for -no-shared-libs

Changes in GPR#1063 weren't applied to nodynlink.ml

dra27 committed:
Add no-shared-libraries to ocamltest

dra27 committed:
Update tool-caml-tex test for -no-shared-libs

dra27 committed:
Fix -compat-32 error message in -custom mode

In -custom mode, then bytecode executable is first made in /tmp and this
filename leaked if the executable fails the -compat-32 test.

dra27 committed:
Fix using -custom with -no-shared-libs

ocamltest links bytecode executables in custom-runtime mode if they have
C files, but it should also do this for any .cma files which require C
stubs on platforms which don't support dynamic C libraries.

dra27 committed:
Skip toplevel tests in -no-shared-libs

If the toplevel doesn't support shared libraries, skip any toplevel
tests which will load cma files which have C stubs (e.g. unix.cma). The
present implementation is suboptimal - really, it would be better to
create a custom toplevel, but it only affects one test at the moment.

dra27 commented on ocamltest/ocaml_actions.ml:
```diff
@@ -952,71 +976,83 @@ let compile_modules
+            what (String.concat " " commandline) exit_status) in
+        (Result.fail_with_reason reason, env)
+      end
+    end else (result, env)
```
The code of the `else` branch has just been updated for indentation

dra27 replied:
(oh, and to use the `backend` variable declared earlier)

damiendoligez replied:
Also `libraries`.

gasche commented:
(I ran another precheck job: https://ci.inria.fr/ocaml/job/precheck/168/ . I'm afraid there is not much else I can do to help. )

dra27 commented:
Thanks @gasche! Perhaps @mshinwell can cast an eye over 4190b3b quickly? a813fe7 just adds `Config.supports_shared_libraries` which should arguably have been there already, so just needs a quick nod.

The remainder of the commits affect the testsuite only, so if @shindere has time to review then great, but equally the changes only affect the tests and do allow us to get CI fully working again.

mshinwell commented:
@dra27 Could you just replace the `nodynlink.ml` changes with the contents of #2197?

shindere commented:
@dra27 yes yes I will review this, it's on my list.
Just a bit overwhelmed these days but will try to do it ASAP.

If you feel I am taking too much time for your taste, don't feel blocked
and just go ahead, I can also review the testsuite part after merge and
submit another PR, at least as far as I am concerned that would be fine
too.

damiendoligez commented on ocamltest/ocaml_actions.ml:
```diff
@@ -177,6 +177,23 @@ let get_program_file backend env =
+      seek_in ic toc_pos;
+      let toc = (input_value ic : Cmo_format.library) in
+      close_in ic;
+      List.compare_length_with toc.Cmo_format.lib_dllibs 0 <> 0
```
This is just a roundabout way of writing `toc.Cmo_format.lib_dllibs <> []`, right?

dra27 replied:
Roundabout is polite!

damiendoligez commented on ocamltest/ocaml_actions.ml:
```diff
@@ -177,6 +177,23 @@ let get_program_file backend env =
+      close_in ic;
+      List.compare_length_with toc.Cmo_format.lib_dllibs 0 <> 0
+    else
+      assert false (* How have non-cma files ended up on the libraries? *)
```
A corrupted or truncated `cma` file would trigger this assertion. Better to output a nice error message. Or maybe just ignore the file: I presume it will be passed to the compiler, which will report the error to the user.

dra27 replied:
I've put the fix for this in a separate commit as it'll want reviewing

damiendoligez commented on ocamltest/ocaml_actions.ml:
```diff
@@ -952,71 +976,83 @@ let compile_modules
-    end
-  end else (result, env)
+  let libraries = libraries toplevel#backend env in
+  let backend = toplevel#backend in
```
Minor nit: you could switch these two definitions and save a call to `toplevel#backend`.

damiendoligez commented on otherlibs/dynlink/nodynlink.ml:
```diff
@@ -19,37 +19,60 @@
-  let iter_initial_units _ = ()
+  let num_globals_inited = not_available
+
+  let fold_initial_units ~init ~f:_ = not_available init
```
The only functional difference with @mshinwell's patch is that `adapt_filename`, `num_globals_inited` and `fold_initial_units` raise `Failure` here instead of returning dummy values. Unless Mark has an objection, I think that's OK.

dra27 replied:
~~I don't follow the "Also `libraries`." comment?~~ GitHub UI being unclear...

damiendoligez commented on utils/config.mli:
```diff
@@ -235,6 +235,11 @@ val flat_float_array : bool
 
+val supports_shared_libraries: bool
+(** Whether shared libraries are supported
+
```
I'd remove the empty line here.

dra27 replied:
FWIW, I added it since `safe_string` and `default_safe_string` are styled that way?

damiendoligez reviewed (approved):
I've reviewed the changes and I have only one real change to suggest (removing the `assert false`). This PR will be good to merge as soon as you deal with that.

shindere commented:
@dra27 do you need some help with that one?

Let me know if you'd like me to propose a rebase

dra27 commented:
Sorry for the lag - I have this rebased and fixed the relevant parts of `configure.ac`, but I have two tests which are unexpectedly failing (I think one will need a trivial update for Cygwin and the other is autoconf-related). I'll push a revised branch once I've fixed those two.

shindere commented:
No problem, David. Let me know if there's anything I can help with.

dra27 committed:
Missing dependency in ocamltest Makefile

dra27 committed:
Propagate --disable-shared correctly on Unix

The --disable-shared option is not presently possible on the native
Windows ports, though this isn't (yet) enforced by configure.

dra27 committed:
Eliminate assert false for corrupt cmas

dra27 commented:
A version of this with the FMA patch is presently going through [precheck](https://ci.inria.fr/ocaml/job/precheck/177/)

dra27 commented:
Tsk, wrong branch in recheck. Now testing the correct code in [precheck#179](https://ci.inria.fr/ocaml/job/precheck/179/).

@shindere / @damiendoligez - has something happened to the mingw32 worker? It was missing the `OCAML_ARCH` variable, which I've restored but it also seems to be missing flexdll (see https://ci.inria.fr/ocaml/job/precheck/179/flambda=false,label=ocaml-mingw-32/console). I have finally uploaded a public key to Inria CI, but I don't think I have access to the workers at the moment?

dra27 commented on configure.ac:
```diff
@@ -365,7 +365,11 @@ AS_IF([test x"$enable_unix_lib" = "xno" -o x"$enable_str_lib" = "xno"],
 # Allow the MSVC linker to be found even if ld isn't installed.
 # User-specified LD still takes precedence.
 AC_CHECK_TOOLS([LD],[ld link])
+# libtool expects host_os=mingw for native Windows
```
Since the autoconf, I've allowed getting `--disable-shared` to work as expected to form part of this GPR. That has also expanded on a problem @shindere identified that `--enable-shared` doesn't work on MSVC (it does work for Cygwin and the mingw-w64 compilers). I did some in-depth poking into the code generated by `LT_INIT` and determined that the problem is that autoconf appears to assume that "native Windows" will be an OS called mingw - there is code in `LT_INIT` which will happily detect the output of `cl`, but it can't work with `$host_os` = `windows`.

This is making me wonder if it would be better for the MSVC ports to be identified not by `pc-windows` but by `pc-mingw32` or `cl-mingw32` or `msvc-mingw32` - i.e. use a different part of the triple to identify it? I don't know - and haven't yet made any effort to find out - what other projects have done about this, but it seems worth determining this now, before `i686-pc-windows` becomes the only way of identifying MSVC.

Thoughts @shindere?

damiendoligez replied:
That sounds fishy. Is autoconf really that ill-adapted to MSVC toolchains? What is $host_os when we run on cygwin?

dra27 replied:
It's not quite the right question - autoconf has reasonable support for `cl` as a compiler, but it does expect a Unix-like OS to be running on ... i.e. it expects the OS to be either `cygwin` or `mingw` (they're just the only two it's aware of). So it's objecting to (or rather, ignoring!) `windows` as the declared OS.

damiendoligez commented:
We fixed the mingw-32 worker, it was just missing the flexdll directory in its PATH.

shindere commented:
Damien Doligez (2019/01/22 06:22 -0800):
> That sounds fishy. Is autoconf really that ill-adapted to MSVC
> toolchains?

Yeah I think they didn't go that far in this direction, due to a lack of
resources to maintain that bit.

> What is $host_os when we run on cygwin?

'cygwin' on both 32 and 64 bits ports.

shindere commented:
David Allsopp (2019/01/15 04:25 -0800):
> Tsk, wrong branch in recheck. Now testing the correct code in
> [precheck#179](https://ci.inria.fr/ocaml/job/precheck/179/).

So, first of all, many thanks for the great work you did!

> @shindere / @damiendoligez - has something happened to the mingw32
> worker? It was missing the `OCAML_ARCH` variable, which I've restored

Thanks a lot for that! I am not sure why the variable was missing, but
as far as I remember, that slave had a problem so was destroyed and
re-created. Most likely, the person who re-created forgot to define the
variable.

> but it also seems to be missing flexdll

As @damiendoligez explained, it was actually installed but not defined
in the PATH when the Jenkins job was started. This is now fixed.

> (see
> https://ci.inria.fr/ocaml/job/precheck/179/flambda=false,label=ocaml-mingw-32/console).
> I have finally uploaded a public key to Inria CI, but I don't think I
> have access to the workers at the moment?

No, indeed, you don't. I'll figure that out and send you a private
e-mail ASAP this works.

shindere commented:
David Allsopp (2019/01/15 20:47 +0000):
> dra27 commented on this pull request.
> 
> 
> 
> > @@ -365,7 +365,11 @@ AS_IF([test x"$enable_unix_lib" = "xno" -o x"$enable_str_lib" = "xno"],
>  # Allow the MSVC linker to be found even if ld isn't installed.
>  # User-specified LD still takes precedence.
>  AC_CHECK_TOOLS([LD],[ld link])
> +# libtool expects host_os=mingw for native Windows
> 
> Since the autoconf, I've allowed getting `--disable-shared` to work as
> expected to form part of this GPR. That has also expanded on a problem
> @shindere identified that `--enable-shared` doesn't work on MSVC (it
> does work for Cygwin and the mingw-w64 compilers). I did some in-depth
> poking into the code generated by `LT_INIT` and determined that the
> problem is that autoconf appears to assume that "native Windows" will
> be an OS called mingw - there is code in `LT_INIT` which will happily
> detect the output of `cl`, but it can't work with `$host_os` =
> `windows`.

Well, many thanks for having clarified this!

> This is making me wonder if it would be better for the MSVC ports to
> be identified not by `pc-windows` but by `pc-mingw32` or `cl-mingw32`
> or `msvc-mingw32` - i.e. use a different part of the triple to
> identify it? I don't know - and haven't yet made any effort to find
> out - what other projects have done about this, but it seems worth
> determining this now, before `i686-pc-windows` becomes the only way of
> identifying MSVC.

@damiendoligez and I have discussed the matter and, at the moment, we
tend to prefer the names we currently have, because we find them
clearer. Also, historically speaking, I started to use libtool because
there were things configure couldn't guess on its own, even on Unix
systems. But then, for Windows, I found myself fixing many of libtool's
guesses to make them fix what we had before. So, recently, I came to
wonder which of libtool's guesses we are still actually using on
Windows. So it's not even obvious to me we should even call LT_INIT on
Windows, in the long run, and that probably will have to be figured out.

All that being said, I don't have storng opinions on this so if you do,
@dra27, there is definitely room for discussion.

damiendoligez commented:
@dra27 Is this good to merge now? I'm waiting for this before I do the beta.
I've relaunched the precheck as https://ci.inria.fr/ocaml/job/precheck/187/ after telling Jenkins to keep precheck builds a little longer...

dra27 commented:
I think my only question was whether we wanted to rename the osname for Windows ... as long as we're happy with the "hack" at the moment for `LT_INIT`, I think this is good to go.

damiendoligez commented:
Perfect, thanks!

Should I squash the commits or do you think it's better to keep the history?

dra27 commented:
Could we keep the history, just for bisection purposes?

damiendoligez commented:
Merged, thanks!

dra27 commented:
Has it been cherry-picked? (not sure from GH interface if it's displaying 4.08 commits in the wrong order, or if they're missing...)
